mistletoe/markdown_renderer.py

```diff
@@ -0,0 +1,537 @@
+        dest (str): link target.
+        title (str): link title (default to empty).
+    """
+
+    repr_attributes = ("label", "dest", "title")
+
+    def __init__(self, match):
+        self.label, self.dest, self.title, self.dest_type, self.title_delimiter = match
+
+
+class LinkReferenceDefinitionBlock(block_token.Footnote):
+    """
+    A sequence of link reference definitions.
+    This is a container block token. Its children are link reference definition tokens.
+
+    This class inherits from Footnote and modifies the behavior of the constructor,
+    to keep the tokens in the AST.
+    """
+
+    def __new__(cls, *args, **kwargs):
+        obj = object.__new__(cls)
+        obj.__init__(*args, **kwargs)
+        return obj
+
+    def __init__(self, matches):
+        self.children = list(map(LinkReferenceDefinition, matches))
+
+
+class Fragment:
+    """
+    Markdown fragment. Used when rendering trees of span tokens into flat sequences.
+    May carry additional data in addition to the text.
+
+    Attributes:
+        text (str): markdown fragment
+    """
+
+    def __init__(self, text: str, **extras):
+        self.text = text
+        for k, v in extras.items():
+            setattr(self, k, v)
+
+
+class MarkdownRenderer(BaseRenderer):
+    """
+    Markdown renderer.
+
+    Designed to make as "clean" a roundtrip as possible, markdown -> parsing -> rendering -> markdown,
+    except for nonessential whitespace. Except when rendering with word wrapping enabled.
+
+    Includes HTMLBlocks and HTMLSpans in the parsing.
+    """
+
+    _whitespace = re.compile(r"\s+")
+
+    def __init__(self, *extras):
+        block_token.remove_token(block_token.Footnote)
+        super().__init__(
+            *chain(
+                (
+                    block_token.HTMLBlock,
+                    span_token.HTMLSpan,
+                    BlankLine,
+                    LinkReferenceDefinitionBlock,
+                ),
+                extras,
+            )
+        )
+        self.render_map["SetextHeading"] = self.render_setext_heading
+        self.render_map["CodeFence"] = self.render_fenced_code_block
+        self.render_map[
+            "LinkReferenceDefinition"
+        ] = self.render_link_reference_definition
+
+    def render(self, token: token.Token, max_line_length: int = None) -> str:
+        """
+        Renders the tree of tokens rooted at the given token into markdown.
+
+        If `max_line_length` is specified, the document is word wrapped to the
+        specified line length. Otherwise the formatting from the original (parsed)
+        document is retained as much as possible.
+        """
+        if isinstance(token, block_token.BlockToken):
+            lines = self.render_map[token.__class__.__name__](
+                token, max_line_length=max_line_length
+            )
+        else:
+            lines = self.span_to_lines([token], max_line_length=max_line_length)
```

Comment:
While it seems it to work, I need to think about some simplifications here. What bothers me is that we have to know / pass _everywhere_ in the nested (render) functions:

1. Whether children are block, or span tokens - i.e. to distinguish calling different (helper) functions. Quick idea: Cannot we generalize this somehow? For example to define a "good old" `render_inner()` which would handle the if-then-else in a similar fashion as can be seen here? Or to have Fragments returned even from blocks rendering functions?

2. The options - now it is just `max_line_length`, but what if there will be more? Quick idea: What about storing the options in `self` (passing it to the constructor already?)?

Reply:
For 1, I think this is less of a problem than what it might seem like. The `render` method is the _only_ method in the class where the check is needed, because in all the rendering methods we already know whether the children are block or span tokens. A token will only have children of one or the other kind.

The `render_inner`-like method you're looking for is probably the dynamic duo consisting of `blocks_to_lines` and `span_to_lines`.

Reply:
For 2, sure, it could become an issue. And I see that some of the other renderers have options in the constructors, so that seems like a reasonable approach.

Reply:
Ad 1 - I agree, technically, it is probably not that big problem, but I'm still unsure if it won't complicate some stuff later on. I would say "will see" for now.

Ad 2 - Yes, storing the options in the constructor is the only feasible way I know of currently. I was also thinking about some globally set-able "mistletoe.options", but making something global seems risky and it would also be behind the scope of this PR I guess. ;)

Reply:
I'll go with the options in the constructors for now.